File: packages/python/elizaos/bootstrap/providers/providers_list.py
from __future__ import annotations

from typing import TYPE_CHECKING

from elizaos.generated.spec_helpers import require_provider_spec
from elizaos.types import Provider, ProviderResult

if TYPE_CHECKING:
    from elizaos.types import IAgentRuntime, Memory, State

# Get text content from centralized specs
_spec = require_provider_spec("PROVIDERS")


async def get_providers_list(
    runtime: IAgentRuntime,
    message: Memory,
    state: State | None = None,
) -> ProviderResult:
    provider_info: list[dict[str, str | bool]] = []

    for provider in runtime.providers:
        provider_info.append(
            {
                "name": provider.name,
                "description": getattr(provider, "description", "No description"),
                "dynamic": getattr(provider, "dynamic", True),
            }
        )

    if not provider_info:
        return ProviderResult(
            text="No providers available.",
            values={"providerCount": 0},
            data={"providers": []},
        )

    formatted_providers = "\n".join(f"- {p['name']}: {p['description']}" for p in provider_info)

    text = f"# Available Providers\n{formatted_providers}"

    return ProviderResult(
        text=text,
        values={
            "providerCount": len(provider_info),
            "providerNames": [p["name"] for p in provider_info],
        },
        data={
            "providers": provider_info,
        },
    )


providers_list_provider = Provider(
    name=_spec["name"],
    description=_spec["description"],
    get=get_providers_list,
    dynamic=_spec.get("dynamic", False),
)

File: packages/rust/src/bootstrap/providers/providers_list.rs
//! PROVIDERS provider implementation.

use async_trait::async_trait;

use crate::error::PluginResult;
use crate::runtime::IAgentRuntime;
use crate::types::{Memory, ProviderResult, State};

use super::Provider;

/// Provider for listing available providers.
pub struct ProvidersListProvider;

#[async_trait]
impl Provider for ProvidersListProvider {
    fn name(&self) -> &'static str {
        "PROVIDERS"
    }

    fn description(&self) -> &'static str {
        "Available context providers"
    }

    fn is_dynamic(&self) -> bool {
        false
    }

    async fn get(
        &self,
        _runtime: &dyn IAgentRuntime,
        _message: &Memory,
        _state: Option<&State>,
    ) -> PluginResult<ProviderResult> {
        // Get providers from the bootstrap plugin itself
        let providers = super::all_providers();

        if providers.is_empty() {
            return Ok(
                ProviderResult::new("No providers available.").with_value("providerCount", 0i64)
            );
        }

        let provider_info: Vec<serde_json::Value> = providers
            .iter()
            .map(|p| {
                serde_json::json!({
                    "name": p.name(),
                    "description": p.description(),
                    "dynamic": p.is_dynamic()
                })
            })
            .collect();

        let formatted: Vec<String> = providers
            .iter()
            .map(|p| format!("- {}: {}", p.name(), p.description()))
            .collect();

        let text = format!("# Available Providers\n{}", formatted.join("\n"));

        let names: Vec<&str> = providers.iter().map(|p| p.name()).collect();

        Ok(ProviderResult::new(text)
            .with_value("providerCount", providers.len() as i64)
            .with_data(
                "providerNames",
                serde_json::to_value(&names).unwrap_or_default(),
            )
            .with_data(
                "providers",
                serde_json::to_value(&provider_info).unwrap_or_default(),
            ))
    }
}
